artichoke-backend/src/extn/core/matchdata/mod.rs

```diff
@@ -308,7 +308,12 @@ impl MatchData {
             } else {
                 haystack.len()
             };
-            let offset = self.region.offset();
+            let offset = self
+                .haystack
```

Comment:
having the `haystack` be stored in the `MatchData` value as a `Vec<u8>` instead of a `spinoso_string::String` is a mistake that will make fixing this bug correctly impossible.

See also this related ticket: https://github.com/artichoke/artichoke/issues/614.

If you're up for it, we'll want to change the `MatchData` rust struct to not store the haystack and regexp within it. Additionally, we should be storing the `Value`s that represent the haystack and regexp as instance variables on the `MatchData`'s `Value` (Setting these `Value`s as ivars is necessary so mruby can correctly mark these stored values during GC due to the FFI internals of the Rust/C `MatchData` bridge. we can chat about this on Discord if you're curious about learning more.)

I think setting ivars on values is not something Artichoke supports right now. This is a missing API in `artichoke-core`: https://artichoke.github.io/artichoke/artichoke_core/.

Reply:
A hack to workaround this gap would be to store the `spinoso_string::String` on the Rust `MatchData` struct instead of a `Vec<u8>`.